dynamiqs/core/propagator_solver.py

```diff
@@ -0,0 +1,57 @@
+from __future__ import annotations
+
+from abc import abstractmethod
+
+import jax
+import jax.numpy as jnp
+from jax import Array
+from jaxtyping import PyTree, Scalar
+
+from ..time_array import ConstantTimeArray
+from .abstract_solver import BaseSolver, MESolver
+
+
+class PropagatorSolver(BaseSolver):
+    def __init__(self, *args):
+        super().__init__(*args)
+        # check that Hamiltonian is time-independent
+        if not isinstance(self.H, ConstantTimeArray):
+            raise TypeError(
+                'Solver `Propagator` requires a time-independent Hamiltonian.'
+            )
+        self.H = self.H.x
+
+    def run(self) -> PyTree:
+        # === solve differential equation
+        def f(y, delta_t):
+            # propagate forward except if delta_t is zero
+            y = jax.lax.cond(delta_t == 0, lambda: y, lambda: self.forward(delta_t, y))
+            # save result
+            res = self.save(y)
```

Comment:
Shouldn't we save only if `delta_t == 0` ?

Reply:
If `tsave = [0, 1, 2, 3]`, then `delta_ts = [0, 1, 1, 1]` (at first iteration we don't propagate and save directly, and for the next three iterations we propagate for `delta_t=1`). I guess this also answers next comment. Once we fix https://linear.app/dynamiqs/issue/DYN-17/add-option-to-evolve-from-t0-=-00 we can do:
```python
delta_ts = jnp.diff(jnp.concatenate((self.options.t0, self.ts)))
```